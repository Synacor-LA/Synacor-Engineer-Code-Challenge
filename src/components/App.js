import React from 'react';
import Weather from './Weather';

/** Fetches the user's latitude and longitude from the weather API
 * @returns { Promise } Promise which resolves to an object like { latitude: #, longitude: # }
 */
const getLatLong = () => {
  return new Promise(resolve => {
    fetch('https://weathersync.herokuapp.com/ip')
      .then(resp => resp.json())
      .then(latLong => resolve(latLong))
      .catch(reason => {
        console.error(reason);
        resolve({ latitude: 34, longitude: -118 });
      });
  });
};

/** Fetches the realtime weather information from the weather API
 * @param { object } object An object with 2 properties: latitude and longitude, both numbers
 * @returns { Promise } Promise which resolves to a weatherData object, see test/dummy.js for example
 */
const getWeatherData = ({ latitude, longitude }) => {
  return new Promise(resolve => {
    fetch(`https://weathersync.herokuapp.com/weather/${latitude},${longitude}`)
      .then(resp => resp.json())
      .then(data => resolve(data))
      .catch(reason => console.error(reason));
  });
};

class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      weatherData: null,
    };
  }

  componentDidMount() {
    getLatLong()
      .then(getWeatherData)
      .then(weatherData => {
        this.setState({ weatherData });
      })
      .catch(reason => console.error(reason));
  }

  render() {
    const { weatherData } = this.state;
    if (!weatherData) return <div id="content">Loading...</div>;
    return (
      <div id="content">
        <h1>CURRENT CONDITIONS FOR</h1>
        <Weather weatherData={weatherData} />
      </div>
    );
  }
}

export default App;
