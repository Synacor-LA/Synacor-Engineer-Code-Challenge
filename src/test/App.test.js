import React from 'react';
import { shallow, mount } from 'enzyme';
import weatherData from './dummy';
import App from '../components/App';

// mock fetch for custom responses
const fetchMock = require('fetch-mock');

fetchMock.get('begin:https://weathersync.herokuapp.com/ip', { latitude: 32, longitude: 118 });
fetchMock.get('begin:https://weathersync.herokuapp.com/weather', weatherData);

describe('App tests', () => {
  it('renders', () => {
    expect.assertions(2);
    const wrapper = shallow(<App />);
    expect(wrapper.exists()).toBe(true);
    expect(wrapper.find('#content')).toHaveLength(1);
  });

  it('calls componentDidMount', () => {
    expect.assertions(1);
    const cdmSpy = spyOn(App.prototype, 'componentDidMount');
    shallow(<App />);
    expect(cdmSpy).toHaveBeenCalledTimes(1);
  });

  it('matches initial snapshot for "loading..." placeholder', () => {
    expect.assertions(1);
    const wrapper = shallow(<App />);
    expect(wrapper).toMatchSnapshot();
  });

  it('makes 2 API calls', done => {
    expect.assertions(1);
    jest.spyOn(global, 'fetch');
    shallow(<App />);
    process.nextTick(() => {
      expect(global.fetch).toHaveBeenCalledTimes(2);
      done();
    });
  });

  it('matches snapshot after API calls', done => {
    expect.assertions(2);
    jest.spyOn(global, 'fetch');
    const wrapper = mount(<App />);
    process.nextTick(() => {
      expect(global.fetch).toHaveBeenCalledTimes(2);
      wrapper.update();
      expect(wrapper).toMatchSnapshot();
      done();
    });
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });
});
