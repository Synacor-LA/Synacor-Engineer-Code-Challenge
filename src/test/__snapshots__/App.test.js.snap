// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App tests matches initial snapshot for "loading..." placeholder 1`] = `
<div
  id="content"
>
  Loading...
</div>
`;

exports[`App tests matches snapshot after API calls 1`] = `
<App>
  <div
    id="content"
  >
    <h1>
      CURRENT CONDITIONS FOR
    </h1>
    <Weather
      weatherData={
        Object {
          "base": "stations",
          "clouds": Object {
            "all": 40,
          },
          "cod": 200,
          "coord": Object {
            "lat": 34,
            "lon": -118.43,
          },
          "dt": 1552267336,
          "id": 5370542,
          "main": Object {
            "humidity": 66,
            "pressure": 1014,
            "temp": 285.74,
            "temp_max": 288.71,
            "temp_min": 282.59,
          },
          "name": "Marina del Rey",
          "rain": Object {
            "1h": 0.96,
          },
          "sys": Object {
            "country": "US",
            "id": 5872,
            "message": 0.0109,
            "sunrise": 1552313376,
            "sunset": 1552355896,
            "type": 1,
          },
          "visibility": 16093,
          "weather": Array [
            Object {
              "description": "light rain",
              "icon": "10n",
              "id": 500,
              "main": "Rain",
            },
          ],
          "wind": Object {
            "deg": 250,
            "speed": 3.1,
          },
        }
      }
    >
      <div
        id="city"
      >
        Marina del Rey
      </div>
      <div
        id="temperature"
      >
        55Â° F
      </div>
      <img
        alt="light rain"
        id="icon"
        src="http://openweathermap.org/img/w/10n.png"
      />
      <div
        id="main-conditions"
      >
        Rain
      </div>
      <p
        id="blurb"
      >
        <strong>
           Marina del Rey 
        </strong>
        is currently experiencing
        <strong>
           light rain
        </strong>
        . It is a cool 55 degrees, with a humidity of 66.
      </p>
    </Weather>
  </div>
</App>
`;
